File: results/2026-07-28_195532/tasks/11-semantic-version-bumper/default-haiku45/generated-code/semantic_version_bumper.py
"""Semantic version bumper based on conventional commits."""
import json
import re
import subprocess
import sys
from pathlib import Path
from typing import Tuple, List


def parse_version(file_path: str) -> str:
    """
    Parse semantic version from package.json.

    Args:
        file_path: Path to package.json

    Returns:
        Version string (e.g., '1.0.0')
    """
    with open(file_path, 'r') as f:
        data = json.load(f)
    return data['version']


def parse_commits(commits: List[str]) -> str:
    """
    Determine version bump type from conventional commit messages.

    Rules:
    - breaking change (feat! or fix!): major
    - feat: minor
    - fix: patch
    - other: no bump

    Args:
        commits: List of commit message strings

    Returns:
        Bump type: 'major', 'minor', 'patch', or 'none'
    """
    has_breaking = False
    has_feat = False
    has_fix = False

    for commit in commits:
        if re.search(r'^(feat|fix)!:', commit):
            has_breaking = True
        elif commit.startswith('feat:'):
            has_feat = True
        elif commit.startswith('fix:'):
            has_fix = True

    if has_breaking:
        return 'major'
    elif has_feat:
        return 'minor'
    elif has_fix:
        return 'patch'
    else:
        return 'none'


def bump_version(current_version: str, bump_type: str) -> str:
    """
    Bump semantic version based on bump type.

    Args:
        current_version: Current version string (e.g., '1.2.3')
        bump_type: Type of bump ('major', 'minor', 'patch', 'none')

    Returns:
        New version string
    """
    parts = current_version.split('.')
    major, minor, patch = int(parts[0]), int(parts[1]), int(parts[2])

    if bump_type == 'major':
        major += 1
        minor = 0
        patch = 0
    elif bump_type == 'minor':
        minor += 1
        patch = 0
    elif bump_type == 'patch':
        patch += 1

    return f'{major}.{minor}.{patch}'


def update_version(file_path: str, new_version: str) -> None:
    """
    Update version in package.json file.

    Args:
        file_path: Path to package.json
        new_version: New version string
    """
    with open(file_path, 'r') as f:
        data = json.load(f)

    data['version'] = new_version

    with open(file_path, 'w') as f:
        json.dump(data, f, indent=2)


def generate_changelog_entry(version: str, commits: List[str]) -> str:
    """
    Generate a changelog entry from commits.

    Args:
        version: Version string
        commits: List of commit messages

    Returns:
        Formatted changelog entry
    """
    entry = f'## [{version}]\n\n'

    features = []
    fixes = []
    other = []

    for commit in commits:
        if commit.startswith('feat'):
            features.append(commit)
        elif commit.startswith('fix'):
            fixes.append(commit)
        else:
            other.append(commit)

    if features:
        entry += '### Features\n'
        for feat in features:
            desc = re.sub(r'^feat!?:\s*', '', feat)
            entry += f'- {desc}\n'
        entry += '\n'

    if fixes:
        entry += '### Bug Fixes\n'
        for fix in fixes:
            desc = re.sub(r'^fix!?:\s*', '', fix)
            entry += f'- {desc}\n'
        entry += '\n'

    return entry


def get_commits_since_tag(repo_path: str, tag: str) -> List[str]:
    """
    Get commit messages since a specific git tag.

    Args:
        repo_path: Path to git repository
        tag: Git tag to use as base

    Returns:
        List of commit messages
    """
    try:
        result = subprocess.run(
            ['git', 'log', f'{tag}..HEAD', '--oneline'],
            cwd=repo_path,
            capture_output=True,
            text=True,
            check=True,
        )
        lines = result.stdout.strip().split('\n')
        commits = [line.split(' ', 1)[1] for line in lines if line]
        return commits
    except subprocess.CalledProcessError as e:
        raise RuntimeError(f'Failed to get commits from git: {e.stderr}') from e

File: results/2026-07-28_180536/tasks/11-semantic-version-bumper/default-haiku45/generated-code/src/semantic-version-bumper.js
const fs = require('fs');
const path = require('path');
const { execSync } = require('child_process');

// Parse semantic version from package.json or VERSION file
function parseVersion(filePath) {
  const ext = path.extname(filePath);

  if (ext === '.json') {
    const content = JSON.parse(fs.readFileSync(filePath, 'utf8'));
    return content.version;
  }

  // For VERSION file or plain text
  return fs.readFileSync(filePath, 'utf8').trim();
}

// Parse conventional commit message to extract type and scope
function parseConventionalCommit(message) {
  // Pattern: type(scope): message or type!: message for breaking changes
  const regex = /^(feat|fix|docs|style|refactor|perf|test|chore|ci)(\(.+\))?(!)?:/;
  const match = message.match(regex);

  if (!match) {
    return null;
  }

  const type = match[1];
  const isBreaking = !!match[3];

  return {
    type,
    isBreaking,
    message: message.substring(match[0].length).trim(),
  };
}

// Get commits since a certain ref with conventional commit format
function getConventionalCommits(sinceRef) {
  try {
    const output = execSync(`git log ${sinceRef}..HEAD --pretty=format:%s`, {
      encoding: 'utf8',
    });

    if (!output.trim()) {
      return [];
    }

    return output
      .split('\n')
      .filter(line => line.trim())
      .map(message => {
        const parsed = parseConventionalCommit(message);
        return parsed || { type: 'chore', message, isBreaking: false };
      })
      .filter(c => c.type !== null);
  } catch (error) {
    // If sinceRef doesn't exist, try all commits
    if (error.message.includes('unknown revision')) {
      const output = execSync('git log --pretty=format:%s', { encoding: 'utf8' });
      return output
        .split('\n')
        .filter(line => line.trim())
        .map(message => {
          const parsed = parseConventionalCommit(message);
          return parsed || { type: 'chore', message, isBreaking: false };
        })
        .filter(c => c.type !== null);
    }
    throw error;
  }
}

// Determine the next version based on commits
function getNextVersion(currentVersion, commits) {
  const [major, minor, patch] = currentVersion.split('.').map(Number);

  // Determine the type of bump needed
  let bumpType = 'none'; // none, patch, minor, major

  for (const commit of commits) {
    // Check for breaking changes (indicated by isBreaking flag or ! in type)
    if (commit.isBreaking) {
      bumpType = 'major';
      break;
    }

    if (commit.type === 'feat' && bumpType !== 'major') {
      bumpType = 'minor';
    }

    if (commit.type === 'fix' && bumpType === 'none') {
      bumpType = 'patch';
    }
  }

  // Calculate new version
  if (bumpType === 'major') {
    return `${major + 1}.0.0`;
  }

  if (bumpType === 'minor') {
    return `${major}.${minor + 1}.0`;
  }

  if (bumpType === 'patch') {
    return `${major}.${minor}.${patch + 1}`;
  }

  return currentVersion;
}

// Update version in package.json or VERSION file
function updateVersionFile(filePath, newVersion) {
  const ext = path.extname(filePath);

  if (ext === '.json') {
    const content = JSON.parse(fs.readFileSync(filePath, 'utf8'));
    content.version = newVersion;
    fs.writeFileSync(filePath, JSON.stringify(content, null, 2) + '\n', 'utf8');
  } else {
    fs.writeFileSync(filePath, `${newVersion}\n`, 'utf8');
  }
}

// Generate changelog entry from commits
function generateChangelog(version, commits) {
  const features = commits.filter(c => c.type === 'feat').map(c => c.message);
  const fixes = commits.filter(c => c.type === 'fix').map(c => c.message);

  let changelog = `## ${version}\n\n`;

  if (features.length > 0) {
    changelog += '### Features\n';
    features.forEach(f => {
      changelog += `- ${f}\n`;
    });
    changelog += '\n';
  }

  if (fixes.length > 0) {
    changelog += '### Bug Fixes\n';
    fixes.forEach(f => {
      changelog += `- ${f}\n`;
    });
    changelog += '\n';
  }

  return changelog;
}

// Main function to bump version
function bumpVersion(versionFilePath, fromRef = 'HEAD~10') {
  try {
    // Parse current version
    const currentVersion = parseVersion(versionFilePath);

    // Get commits since ref
    const commits = getConventionalCommits(fromRef);

    if (commits.length === 0) {
      return {
        currentVersion,
        newVersion: currentVersion,
        commits: [],
        changelog: 'No commits to process\n',
      };
    }

    // Calculate next version
    const newVersion = getNextVersion(currentVersion, commits);

    // Update version file
    updateVersionFile(versionFilePath, newVersion);

    // Generate changelog
    const changelog = generateChangelog(newVersion, commits);

    return {
      currentVersion,
      newVersion,
      commits,
      changelog,
    };
  } catch (error) {
    throw new Error(`Failed to bump version: ${error.message}`);
  }
}

module.exports = {
  parseVersion,
  getNextVersion,
  updateVersionFile,
  generateChangelog,
  getConventionalCommits,
  bumpVersion,
};
